### PR TITLE
Improve snapshot load error handling.

### PR DESCRIPTION
**Pull request by sandreim**
## Reason for This PR

Improves snapshot load error handling by adding extra checks to ensure data or format versions are valid before attempting to deserialize the snapshot.

## Description of Changes

Add snapshot version errors and checks.
    
    - implement checks for snapshot format/data version
    - add negative unit tests

- [ ] This functionality can be added in [`rust-vmm`](https://github.com/rust-vmm).

## License Acceptance

By submitting this pull request, I confirm that my contribution is made under
the terms of the Apache 2.0 license.

## PR Checklist

`[Author TODO: Meet these criteria.]`
`[Reviewer TODO: Verify that these criteria are met. Request changes if not]`

- [x] All commits in this PR are signed (`git commit -s`).
- [x] The reason for this PR is clearly provided (issue no. or explanation).
- [x] The description of changes is clear and encompassing.
- [x] Any required documentation changes (code and docs) are included in this PR.
- [x] Any newly added `unsafe` code is properly documented.
- [x] Any API changes are reflected in `firecracker/swagger.yaml`.
- [x] Any user-facing changes are mentioned in `CHANGELOG.md`.
- [x] All added/changed functionality is tested.

### CODE REVIEW

File: src/snapshot/tests/test.rs [hunk 1]
```diff
@@ -116,3 +116,115 @@ fn test_hardcoded_snapshot_deserialization() {
...
+        0x01, 0x00,
+        // `a` field +
+        0x10, 0x00, 0x00, 0x00,
+        // `b` field: Option variant type (1 byte) + inner enum variant type (4 bytes)
+        // + inner enum value (4 bytes).
+        0x01, 0x01, 0x00, 0x00, 0x00, 0x02, 0x00, 0x00, 0x00,
+    ];
+
+    let mut snapshot_blob = invalid_format_snap;
+    let mut result: Result<A, Error> = Snapshot::load(&mut snapshot_blob, VersionMap::new());
```
Comment (georgepisaltu) [inline src/snapshot/tests/test.rs:145]: You could drop `snapshot_blob` here as well as in the next snapshot loads and just pass the reference to the bytes.

```suggestion
    let mut result: Result<A, Error> = Snapshot::load(&mut invalid_format_snap.as_ref(), VersionMap::new());
```

File: src/snapshot/tests/test.rs [hunk 2]
```diff
@@ -116,3 +116,115 @@ fn test_hardcoded_snapshot_deserialization() {
 
     assert_eq!(restored_struct, expected_struct);
 }
+
+#[test]
+fn test_invalid_format_version() {
+    #[rustfmt::skip]
+    let invalid_format_snap: &[u8] = &[
+        // This blob is consisted of the following: magic_id (8 bytes),
```
Comment (georgepisaltu) [inline src/snapshot/tests/test.rs:124]: Nit, same for next occurrences.

```suggestion
        // This blob consists of the following: magic_id (8 bytes),
```

File: src/snapshot/tests/test.rs (hunk 1 above)
Comment (sandreim) [reply src/snapshot/tests/test.rs:145]: done

File: src/snapshot/tests/test.rs (hunk 2 above)
Comment (sandreim) [reply src/snapshot/tests/test.rs]: done

**Comment by sandreim** [comment]
@acatangiu @georgepisaltu I've updated this PR. PTAL!

**Review by acatangiu — APPROVED** [review]

**Review by georgepisaltu — APPROVED** [review]

**Pull request merged**